pallene/constant_propagation.lua

```diff
@@ -0,0 +1,131 @@
+        while #stack > 0 do
+            local cmd = table.remove(stack)
+            local tag = cmd._tag
+            if     tag == 'ir.Cmd.SetGlobal' then
+                if is_constant_value(cmd.src) then
+                    constant_initializer[cmd.global_id] = cmd.src
+                end
+            elseif tag == 'ir.Cmd.Seq' then
+                for i = #cmd.cmds, 1, -1 do
+                    table.insert(stack, cmd.cmds[i])
+                end
+            else
+                -- skip
+            end
+        end
+    end
+
+    -- 2) Find what toplevel variables are never re-initialized or never used
+
+    local n_reads  = {} -- { g_id => int }
+    local n_writes = {} -- { g_id => int }
+    for i = 1, n_globals do
+        n_reads[i]  = 0
+        n_writes[i] = 0
+    end
+
+    for _, func in ipairs(module.functions) do
+        for _, cmd in ipairs(ir.flatten_cmd(func.body)) do
+            local tag = cmd._tag
+            if     tag == "ir.Cmd.GetGlobal" then
+                local id = cmd.global_id
+                n_reads[id] = n_reads[id] + 1
+            elseif tag == "ir.Cmd.SetGlobal" then
+                local id = cmd.global_id
+                n_writes[id] = n_writes[id] + 1
+            else
+                -- skip
+            end
+        end
+    end
+
+    -- 3) Find out which constant globals should be propagated, and which
+    --    unused constant globals should be simply eliminated.
+
+    local n_new_globals = 0
+    local new_global_id = {} -- { g_id => g_id? }
+    for i = 1, n_globals do
+        if constant_initializer[i] and (n_reads[i] == 0 or n_writes[i] == 1) then
+            new_global_id[i] = false
+        else
+            n_new_globals = n_new_globals + 1
+            new_global_id[i] = n_new_globals
+        end
+    end
+
+    -- 4) Propagate the constant globals, and rename the existing ones
+    -- accordingly
+
+    for _, func in ipairs(module.functions) do
+        func.body = ir.map_cmd(func.body, function(cmd)
+            local tag = cmd._tag
+            if     tag == "ir.Cmd.GetGlobal" then
+                local old_id = cmd.global_id
+                local new_id = new_global_id[old_id]
+                if new_id then
+                    return ir.Cmd.GetGlobal(cmd.loc, cmd.dst, new_id)
+                else
+                    local v = assert(constant_initializer[old_id])
+                    return ir.Cmd.Move(cmd.loc, cmd.dst, v)
+                end
+            elseif tag == "ir.Cmd.SetGlobal" then
+                local old_id = cmd.global_id
+                local new_id = new_global_id[old_id]
+                if new_id then
+                    return ir.Cmd.SetGlobal(cmd.loc, new_id, cmd.src)
+                else
+                    return ir.Cmd.Nop()
+                end
+            else
+                -- don't modify
+                return false
+            end
+        end)
+    end
+
+    -- 5) Done
```

Comment:
I would remove this one :)

Reply:
Can I keep the `5)`? My intention was to clarify that the `ir.clean_all(module)` does not belong under section `4)`